home/.chezmoiscripts/universal/run_onchange_before_11-join-etcd-cluster.tmpl

```diff
@@ -0,0 +1,38 @@
+{{- if (eq .host.distro.family "linux") -}}
+#!/usr/bin/env bash
+# @file etcd Clustering
+# @brief Adds the host to `etcd` cluster
+# @description
+#     This script adds the host to `etcd` cluster provided as input, if `etcd` is installed. Static clustering
+#     mechanism is used to join the host to an existing etcd cluster. The other mechanisms that are based on
+#     service discovery is not supported by this script.
+#
+#     Static mechanism needs the IP address(es) of all the participating members to bootstrap a node/member.
+#     This is the only input needed to be able to add the host as a member of the cluster. The format of this field
+#     is available in the [inputs file](https://github.com/megabyte-labs/install.doctor/blob/master/home/.chezmoi.yaml.tmpl).
+#
+#     ## Configuration Variables
+#
+#     The following chart details the input variable(s) that are used to join the host to `etcd` cluster:
+#
+#     | Variable                     | Description                                                |
+#     |------------------------------|------------------------------------------------------------|
+#     | `etcd.initial_cluster_urls`  | Appropriately formatted initial cluster configuration      |
+
+{{ includeTemplate "universal/profile-before" }}
+{{ includeTemplate "universal/logg-before" }}
+
+### Join host to `etcd` cluster
+if command -v etcd > /dev/null; then
+  logg info 'Adding host to the `etcd` cluster'
+  {{ if and .etcd.initial_cluster_urls (gt (len .etcd.initial_cluster_urls) 0) }}
+  etcd --name {{ .chezmoi.hostname }} \
```

Comment:
I don't think .chezmoi.hostname will do it... it has to be .user.hostname etc. Also, I just tested `hostname -i` and that does not work on Linux. Can we listen to {{ hostname }}.{{ public_services_domain }} instead?

Also, we need to include logic to set up the initial etcd which we'll be doing with Kubernetes so you don't have to worry about it for this script.

How does it authenticate though?

Reply:
Sure, the variable that will always return a value has to be used here.

I tested the `hostname -i` on linux where I setup the cluster. We can use something that does not rely on a command. As long as `{{ hostname }}.{{ public_services_domain }}` will resolve to an accessible IP externally and among the cluster members, this can be used.

The assumption with this script is that the command `etcdctl member add <name> --peer-urls=http://<IP/FQDN>:2380` is executed to add the member to the cluster. Then the `advertise peer URL` is used to start this member. In other words, a cluster should be up and running before this script is run.

Authentication is not added at this point. Username/password can be used. Certificate based auth can also be used. I believe the 2nd one may be better.

Reply:
Hey @enggnr --- let's set up certificate based by storing the certificates in encrypted form inside `.chezmoitemplates/files`. Also, let's get rid of the need for hardcoding any IP addresses. I mentioned how we can do this in other comments --- we can rely on Tailscale or CloudFlare to set up a private network that we can automatically join up to each other for all these services via a broadcast message.

Reply:
Support for TLS is added. The files containing the certs in `.chezmoitemplates` are all the same; added to give an idea.